node/pkg/guardiansigner/amazonkms.go

```diff
@@ -0,0 +1,255 @@
+	ethcrypto "github.com/ethereum/go-ethereum/crypto"
+)
+
+var (
+	secp256k1N     = ethcrypto.S256().Params().N
+	secp256k1HalfN = new(big.Int).Div(secp256k1N, big.NewInt(2))
+
+	// The timeout for KMS operations. This is necessary to avoid situations where
+	// the signing or verification is blocked indefinitely.
+	KMS_TIMEOUT = time.Second * 15
+)
+
+// The ASN.1 structure for an ECDSA signature produced by AWS KMS.
+type asn1EcSig struct {
+	R asn1.RawValue
+	S asn1.RawValue
+}
+
+// The ASN.1 structure for an ECDSA public key produced by AWS KMS.
+type asn1EcPublicKey struct {
+	EcPublicKeyInfo asn1EcPublicKeyInfo
+	PublicKey       asn1.BitString
+}
+
+// The ASN.1 structure for the public key info in an ECDSA public key produced by AWS KMS.
+type asn1EcPublicKeyInfo struct {
+	Algorithm  asn1.ObjectIdentifier
+	Parameters asn1.ObjectIdentifier
+}
+
+// getRegionFromArn extracts the region from an ARN. The region is at index 3 in the ARN.
+func getRegionFromArn(arn string) string {
+	// Information in ARNs are colon-separated
+	arn_parts := strings.Split(arn, ":")
+
+	// https://docs.aws.amazon.com/IAM/latest/UserGuide/reference-arns.html#arns-syntax
+	// The format of an ARN is arn:partition:service:region:account-id:resource-info, so
+	// the region is at index 3.
+	if len(arn) < 4 {
+		return ""
+	}
+
+	return arn_parts[3]
+}
+
+// AmazonKms is a signer that uses AWS KMS to sign messages. The URI is expected to be
+// in the format amazonkms://<key-arn>.
+type AmazonKms struct {
+	keyId     string
+	region    string
+	publicKey ecdsa.PublicKey
+	client    *kms.Client
+}
+
+// NewAmazonKmsSigner creates a new AmazonKms signer. The keyPath is expected to be an ARN,
+// identifying the key in AWS KMS. The region is extracted from the ARN, and the AWS KMS
+// client is created with the region.
+// NOTE: The public key is retrieved during signer creation, and stored as a property of the
+// signer. This is because the public key is not expected to change during runtime.
+func NewAmazonKmsSigner(ctx context.Context, unsafeDevMode bool, keyPath string) (*AmazonKms, error) {
+	timeoutCtx, cancel := context.WithTimeout(ctx, KMS_TIMEOUT)
+	defer cancel()
+
+	// Extract the region from the key path. The region is required to create a new KMS client.
+	// If the region is not present in the key path, the ARN is considered invalid.
+	region := getRegionFromArn(keyPath)
+
+	if region == "" {
+		return nil, errors.New("Invalid KMS ARN")
+	}
+
+	amazonKmsSigner := AmazonKms{
+		keyId:  keyPath,
+		region: region,
+	}
+
+	// Create a configuration object to create a new KMS client from. The region passed to
+	// `config.WithDefaultRegion()` must match the region in the actual ARN, otherwise the SDK throws
+	// an error. This is why the region is first extracted from the keyPath.
+	cfg, err := config.LoadDefaultConfig(timeoutCtx, config.WithDefaultRegion(amazonKmsSigner.region))
+	if err != nil {
+		return nil, errors.New("Failed to load default config")
```

Comment:
Should this error be more descriptive re being related to KMS? i.e. should every error string in this signer include the string "KMS"?